### PR TITLE
[Enhancement] Enhance training process

### PR DESCRIPTION
**Pull request by egpivo**
## Objective

- Enhance training process
## Changes

1. `trainer.py`
   - Class Structure: The DiffusionTrainer class is well-structured and organized. Each method has a clear responsibility, making the code easy to understand.
   - Logging:  Used a logger to handle logging messages, which is a good practice for maintaining logs during training.
   - Type Hints: Type hints are added to function arguments and return values, improving code clarity and aiding in understanding.
  - Error Handling: Proper error handling is implemented, especially in loading checkpoints and validating masking parameters.
   - Readability: The code is clear and readable, making it easier for other developers to comprehend and modify.
2. `train.py`
  - Argument Parsing: Command-line argument parsing is implemented using argparse, allowing for easy configuration of the training process.
  - Configuration Handling: The script reads configuration from a JSON file, making it flexible to adjust various parameters without modifying the code.
  - Model Setup: The script sets up the model according to the configuration, improving modularity and reusability.
  - Output Directory Setup: The output directory is properly set up based on the configuration, ensuring organized storage of checkpoints and logs.
  - Main Function: The main function orchestrates the training process, handling device selection, configuration loading, model setup, and running the training job.

3. Update the Makefile by adding command `train-mix`
## Example

- Use Makefile to train a model
```bash
make train-mix
```

### CODE REVIEW

**Comment by egpivo** [comment]
Check if different floating precisions on MacOs and Ubuntu cause the `Nan` regarding the following script

```python
import sys
import math

# Check machine epsilon and float info
print(f"Machine epsilon: {sys.float_info.epsilon}")
print(f"Float info: {sys.float_info}")

# Perform arithmetic operations
a = 0.1 + 0.2
b = 0.3

# Compare using a tolerance
tolerance = 1e-9
if math.isclose(a, b, rel_tol=tolerance):
    print("The numbers are close enough.")
else:
    print("The numbers are not close enough.")

# Print the actual values
print(f"a: {a}")
print(f"b: {b}")
````

- The results return identically as
```
Machine epsilon: 2.220446049250313e-16
Float info: sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)
The numbers are close enough.
a: 0.30000000000000004
b: 0.3
```

So it can't be the precision.

**Comment by egpivo** [comment]
Parallelism

```
import torch

# Set the number of threads to 1 for both OpenMP and MKL
torch.set_num_threads(1)
```
Set up only single CPU and run on Ubuntu OS. Still got nan.

**Comment by egpivo** [comment]
Most parameters in residual_layer are huge that cause the gradient exploding

```
module.residual_layer.residual_blocks.8.S41.s4_layer.output_linear.bias grad norm: inf
module.residual_layer.residual_blocks.8.S41.s4_layer.output_linear.bias grad norm: inf
module.residual_layer.residual_blocks.8.S41.s4_layer.output_linear.weight grad norm: inf
module.residual_layer.residual_blocks.8.S41.s4_layer.output_linear.weight grad norm: inf
module.residual_layer.residual_blocks.8.S41.s4_layer.D grad norm: inf
module.residual_layer.residual_blocks.8.S41.s4_layer.D grad norm: inf
module.residual_layer.residual_blocks.8.S41.s4_layer.kernel.kernel.C grad norm: inf
module.residual_layer.residual_blocks.8.S41.s4_layer.kernel.kernel.C grad norm: inf
module.residual_layer.residual_blocks.8.conv_layer.conv.bias grad norm: inf
module.residual_layer.residual_blocks.8.conv_layer.conv.bias grad norm: inf
module.residual_layer.residual_blocks.8.conv_layer.conv.parametrizations.weight.original1 grad norm: inf
module.residual_layer.residual_blocks.8.conv_layer.conv.parametrizations.weight.original1 grad norm: inf
module.residual_layer.residual_blocks.8.conv_layer.conv.parametrizations.weight.original0 grad norm: inf
module.residual_layer.residual_blocks.8.conv_layer.conv.parametrizations.weight.original0 grad norm: inf
module.residual_layer.residual_blocks.8.fc_t.bias grad norm: inf
module.residual_layer.residual_blocks.8.fc_t.bias grad norm: inf
module.residual_layer.residual_blocks.8.fc_t.weight grad norm: inf
module.residual_layer.residual_blocks.8.fc_t.weight grad norm: inf
module.residual_layer.residual_blocks.7.skip_conv.bias grad norm: 2.177200985897798e-05
module.residual_layer.residual_blocks.7.skip_conv.bias grad norm: 2.177200985897798e-05
module.residual_layer.residual_blocks.7.skip_conv.parametrizations.weight.original1 grad norm: 0.0002756942412815988
module.residual_layer.residual_blocks.7.skip_conv.parametrizations.weight.original1 grad norm: 0.0002756942412815988
module.residual_layer.residual_blocks.7.skip_conv.parametrizations.weight.original0 grad norm: 1.6145633708219975e-05
module.residual_layer.residual_blocks.7.skip_conv.parametrizations.weight.original0 grad norm: 1.6145633708219975e-05
module.residual_layer.residual_blocks.7.res_conv.bias grad norm: inf
module.residual_layer.residual_blocks.7.res_conv.bias grad norm: inf
module.residual_layer.residual_blocks.7.res_conv.parametrizations.weight.original1 grad norm: inf
module.residual_layer.residual_blocks.7.res_conv.parametrizations.weight.original1 grad norm: inf
module.residual_layer.residual_blocks.7.res_conv.parametrizations.weight.original0 grad norm: inf
module.residual_layer.residual_blocks.7.res_conv.parametrizations.weight.original0 grad norm: inf
module.residual_layer.residual_blocks.7.S42.norm_layer.weight grad norm: inf
module.residual_layer.residual_blocks.7.S42.norm_layer.weight grad norm: inf
module.residual_layer.residual_blocks.7.S42.norm_layer.bias grad norm: inf
module.residual_layer.residual_blocks.7.S42.norm_layer.bias grad norm: inf
module.residual_layer.residual_blocks.7.S42.s4_layer.output_linear.bias grad norm: inf
module.residual_layer.residual_blocks.7.S42.s4_layer.output_linear.bias grad norm: inf
module.residual_layer.residual_blocks.7.S42.s4_layer.output_linear.weight grad norm: inf
module.residual_layer.residual_blocks.7.S42.s4_layer.output_linear.weight grad norm: inf
```
